### PR TITLE
[SPARK-27356][SQL] File source V2: Fix the case that data columns overlap with partition schema

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

In the current file source V2 framework, the schema of `FileScan` is not returned correctly if there are overlap columns between `dataSchema` and `partitionSchema`. The actual schema should be 
`dataSchema - overlapSchema + partitionSchema`, which might have different column order from the pushed down `requiredSchema` in `SupportsPushDownRequiredColumns.pruneColumns`.

For example, if the data schema is `[a: String, b: String, c: String]` and the partition schema is `[b: Int, d: Int]`, the result schema is `[a: String, b: Int, c: String, d: Int]` in current `FileTable` and `HadoopFsRelation`. while the actual scan schema is `[a: String, c: String, b: Int, d: Int]` in `FileScan`.

To fix the corner case, this PR proposes that the output schema of `FileTable` should be `dataSchema - overlapSchema + partitionSchema`, so that the column order is consistent with `FileScan`.
Putting all the partition columns to the end of table schema is more reasonable. 

## How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by gengliangwang** [comment]
@cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala [hunk 1]
```diff
@@ -40,7 +45,23 @@ abstract class FileScan(
...
     val maxSplitBytes = FilePartition.maxSplitBytes(sparkSession, selectedPartitions)
+    val partitionAttributes = fileIndex.partitionSchema.toAttributes
+    val attributeMap = partitionAttributes.map(a => getAttributeName(a) -> a).toMap
+    val readPartitionAttributes = readPartitionSchema.toAttributes.map { readAttr =>
+      attributeMap.get(getAttributeName(readAttr)).getOrElse {
+        throw new AnalysisException(s"Can't find required partition column ${readAttr.name} " +
+          s"in partition schema ${fileIndex.partitionSchema}")
+      }
+    }
+    lazy val partitionValueProject =
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: If this PR contains too many changes, I am OK to create a separate PR for the partition value pruning.

**Comment by gengliangwang** [comment]
retest this please.

File: docs/sql-migration-guide-upgrade.md [hunk 2]
```diff
@@ -50,6 +50,8 @@ license: |
 
   - In Spark version 2.4 and earlier, JSON datasource and JSON functions like `from_json` convert a bad JSON record to a row with all `null`s in the PERMISSIVE mode when specified schema is `StructType`. Since Spark 3.0, the returned row can contain non-`null` fields if some of JSON column values were parsed and converted to desired types successfully.
 
+  - In Spark version 2.4 and earlier, if data columns overlap with partition columns, the output schema of file scan respects the ordering of data columns, and adopts the data type of partition columns. Since Spark 3.0, the output schema of file scan puts all the partition columns at the end. For example, if the data schema is `[a: String, b: String, c: String]` and the partition schema is `[b: Int, d: Int]`, the result schema is `[a: String, b: Int, c: String, d: Int]` in Spark 2.4 and earlier, and `[a: String, c: String, b: Int, d: Int]` since Spark 3.0.
```
Comment (cloud-fan) [inline docs/sql-migration-guide-upgrade.md]: do we need migration guide? it's a behavior change for file source v2, which is new in Spark 3.0.

Comment (gengliangwang) [reply docs/sql-migration-guide-upgrade.md]: I am OK with either way. Let me remove this.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala [hunk 3]
```diff
@@ -127,7 +189,7 @@ abstract class OrcPartitionDiscoveryTest extends OrcTest {
             i <- 1 to 10
             pi <- Seq(1, 2)
             ps <- Seq("foo", "bar")
-          } yield Row(i, pi, i.toString, ps))
+          } yield Row(i, i.toString, pi, ps))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala]: are we testing v1 or v2 here?

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala]: V2.
For V1 we use `OrcV1PartitionDiscoverySuite`.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala]: maybe we should put `V2` in the test suite name as well.

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala]: This is not quite related to this PR. If we are going to use V2 by default, I think the current test suite name is OK.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala]: isn't V1 by default now?

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala]: For read path, it is V2 by default now.

File: docs/sql-migration-guide-upgrade.md [hunk 4]
```diff
@@ -49,7 +49,7 @@ license: |
   - In Spark version 2.4 and earlier, CSV datasource converts a malformed CSV string to a row with all `null`s in the PERMISSIVE mode. Since Spark 3.0, the returned row can contain non-`null` fields if some of CSV column values were parsed and converted to desired types successfully.
 
   - In Spark version 2.4 and earlier, JSON datasource and JSON functions like `from_json` convert a bad JSON record to a row with all `null`s in the PERMISSIVE mode when specified schema is `StructType`. Since Spark 3.0, the returned row can contain non-`null` fields if some of JSON column values were parsed and converted to desired types successfully.
-
+
```
Comment (cloud-fan) [inline docs/sql-migration-guide-upgrade.md]: unnecessary change

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala]: let's make a followup PR to put V2 in the test suite name and do not rely on the default config values.

**Comment by cloud-fan** [comment]
thanks, merging to master!
